File: frontend/src/lib/reviews.ts
import { apiClient, API_ENDPOINTS } from './api';

// 백엔드에서 받아오는 실제 데이터 구조
export interface ReviewResponse {
  id: number;
  user_id: number;
  counselor_id?: number;
  consultation_id?: number;
  rating: number;
  title: string;
  content: string;
  is_approved: boolean;
  is_active: boolean;
  view_count: number;
  author_name?: string;
  counselor_name?: string;
  created_at: string;
  updated_at?: string;
}

// 프론트엔드에서 사용하는 구조
export interface Review {
  id: string;
  rating: number;
  title: string;
  content: string;
  counselor_id?: string;
  consultation_id?: string;
  is_approved: boolean;
  is_active: boolean;
  view_count: number;
  author_name?: string;
  counselor_name?: string;
  created_at: string;
  updated_at?: string;
}

export interface CreateReviewRequest {
  counselor_id?: string;
  consultation_id?: string;
  rating: number;
  title: string;
  content: string;
}

export interface UpdateReviewRequest extends Partial<CreateReviewRequest> {
  id: string;
  is_approved?: boolean;
}

// 백엔드 데이터를 프론트엔드 구조로 변환
function mapReviewResponse(response: ReviewResponse): Review {
  return {
    id: response.id.toString(),
    rating: response.rating,
    title: response.title,
    content: response.content,
    counselor_id: response.counselor_id?.toString(),
    consultation_id: response.consultation_id?.toString(),
    is_approved: response.is_approved,
    is_active: response.is_active,
    view_count: response.view_count || 0,
    author_name: response.author_name,
    counselor_name: response.counselor_name,
    created_at: response.created_at,
    updated_at: response.updated_at
  };
}

// 승인된 후기 목록 가져오기
export async function getApprovedReviews(counselorId?: string): Promise<Review[]> {
  try {
    const endpoint = counselorId 
      ? `${API_ENDPOINTS.REVIEWS}/approved?counselor_id=${counselorId}`
      : `${API_ENDPOINTS.REVIEWS}/approved`;
    
    const response = await apiClient.get<{reviews: ReviewResponse[], total: number, page: number, size: number}>(endpoint);
    const reviews = response.data?.reviews || [];
    return reviews.map(mapReviewResponse);
  } catch (error) {
    console.error('후기 목록을 가져오는데 실패했습니다:', error);
    return [];
  }
}

// 후기 작성
export async function createReview(data: CreateReviewRequest): Promise<Review | null> {
  try {
    const response = await apiClient.post<ReviewResponse>(API_ENDPOINTS.REVIEWS, data);
    return response.data ? mapReviewResponse(response.data) : null;
  } catch (error) {
    console.error('후기 작성에 실패했습니다:', error);
    return null;
  }
}

// 특정 후기 정보 가져오기
export async function getReview(id: string): Promise<Review | null> {
  try {
    const response = await apiClient.get<ReviewResponse>(API_ENDPOINTS.REVIEW(id));
    return response.data ? mapReviewResponse(response.data) : null;
  } catch (error) {
    console.error('후기 정보를 가져오는데 실패했습니다:', error);
    return null;
  }
}

// 후기 수정
export async function updateReview(data: UpdateReviewRequest): Promise<Review | null> {
  try {
    const response = await apiClient.put<ReviewResponse>(API_ENDPOINTS.REVIEW(data.id), data);
    return response.data ? mapReviewResponse(response.data) : null;
  } catch (error) {
    console.error('후기 수정에 실패했습니다:', error);
    return null;
  }
}

// 후기 삭제
export async function deleteReview(id: string): Promise<boolean> {
  try {
    await apiClient.delete(API_ENDPOINTS.REVIEW(id));
    return true;
  } catch (error) {
    console.error('후기 삭제에 실패했습니다:', error);
    return false;
  }
}

File: backend/fastapi-legacy/app/api/reviews.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from ..database import get_db
from ..models.review import Review
from ..models.user import User
from ..schemas.review import ReviewCreate, ReviewUpdate, Review as ReviewSchema, ReviewList
from ..dependencies import get_current_active_user, get_current_admin_user

router = APIRouter(prefix="/reviews", tags=["후기"])


@router.post("/", response_model=ReviewSchema)
def create_review(
    review_data: ReviewCreate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """후기 작성"""
    # 별점 검증 (1-5)
    if not 1 <= review_data.rating <= 5:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="별점은 1-5 사이의 값이어야 합니다."
        )
    
    db_review = Review(
        user_id=current_user.id,
        **review_data.dict()
    )
    
    db.add(db_review)
    db.commit()
    db.refresh(db_review)
    
    return db_review


@router.get("/", response_model=ReviewList)
def get_reviews(
    skip: int = Query(0, ge=0),
    limit: int = Query(10, ge=1, le=100),
    counselor_id: Optional[int] = None,
    is_approved: Optional[bool] = None,
    is_active: Optional[bool] = None,
    db: Session = Depends(get_db)
):
    """후기 목록 조회"""
    query = db.query(Review)
    
    if counselor_id:
        query = query.filter(Review.counselor_id == counselor_id)
    
    if is_approved is not None:
        query = query.filter(Review.is_approved == is_approved)
    
    if is_active is not None:
        query = query.filter(Review.is_active == is_active)
    
    total = query.count()
    reviews = query.offset(skip).limit(limit).all()
    
    return ReviewList(
        reviews=reviews,
        total=total,
        page=skip // limit + 1,
        size=limit
    )


@router.get("/approved", response_model=ReviewList)
def get_approved_reviews(
    skip: int = Query(0, ge=0),
    limit: int = Query(10, ge=1, le=100),
    counselor_id: Optional[int] = None,
    db: Session = Depends(get_db)
):
    """승인된 후기 목록 조회"""
    query = db.query(Review).filter(Review.is_approved == True, Review.is_active == True)
    
    if counselor_id:
        query = query.filter(Review.counselor_id == counselor_id)
    
    total = query.count()
    reviews = query.offset(skip).limit(limit).all()
    
    # 작성자와 상담사 정보 추가
    for review in reviews:
        # 작성자 정보 (익명 처리)
        if review.user and not review.is_anonymous:
            author_name = review.user.full_name
            if author_name:
                # 이름을 익명 처리 (예: "홍길동" -> "홍**")
                if len(author_name) >= 2:
                    review.author_name = author_name[0] + "**"
                else:
                    review.author_name = author_name + "**"
        else:
            review.author_name = "익명"
        
        # 상담사 정보
        if review.counselor:
            review.counselor_name = review.counselor.name
    
    return ReviewList(
        reviews=reviews,
        total=total,
        page=skip // limit + 1,
        size=limit
    )


@router.get("/{review_id}", response_model=ReviewSchema)
def get_review(
    review_id: int,
    db: Session = Depends(get_db)
):
    """후기 상세 조회 (승인된 후기만)"""
    review = db.query(Review).filter(
        Review.id == review_id,
        Review.is_approved == True,
        Review.is_active == True
    ).first()
    
    if not review:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="후기를 찾을 수 없습니다."
        )
    
    # 작성자와 상담사 정보 추가
    # 작성자 정보 (익명 처리)
    if review.user and not review.is_anonymous:
        author_name = review.user.full_name
        if author_name:
            # 이름을 익명 처리 (예: "홍길동" -> "홍**")
            if len(author_name) >= 2:
                review.author_name = author_name[0] + "**"
            else:
                review.author_name = author_name + "**"
    else:
        review.author_name = "익명"
    
    # 상담사 정보
    if review.counselor:
        review.counselor_name = review.counselor.name
    
    return review


@router.put("/{review_id}", response_model=ReviewSchema)
def update_review(
    review_id: int,
    review_update: ReviewUpdate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """후기 수정"""
    review = db.query(Review).filter(Review.id == review_id).first()
    
    if not review:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="후기를 찾을 수 없습니다."
        )
    
    # 본인의 후기이거나 관리자인 경우만 수정 가능
    if review.user_id != current_user.id and not current_user.is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="접근 권한이 없습니다."
        )
    
    # 일반 사용자는 승인 상태를 수정할 수 없음
    if not current_user.is_admin and review_update.is_approved is not None:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="승인 상태는 관리자만 수정할 수 있습니다."
        )
    
    update_data = review_update.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(review, field, value)
    
    db.commit()
    db.refresh(review)
    
    return review


@router.delete("/{review_id}")
def delete_review(
    review_id: int,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """후기 삭제"""
    review = db.query(Review).filter(Review.id == review_id).first()
    
    if not review:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="후기를 찾을 수 없습니다."
        )
    
    # 본인의 후기이거나 관리자인 경우만 삭제 가능
    if review.user_id != current_user.id and not current_user.is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="접근 권한이 없습니다."
        )
    
    db.delete(review)
    db.commit()
    
    return {"message": "후기가 삭제되었습니다."} 